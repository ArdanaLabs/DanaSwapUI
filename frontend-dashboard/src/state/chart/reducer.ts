import {
  ActionReducerMapBuilder,
  PayloadAction,
  createReducer,
} from "@reduxjs/toolkit"

import * as E from "fp-ts/Either"
import { RemoteData, fromEither, pending } from "fp-ts-remote-data"

import {
  receivedAggVolume,
  receivedAggLiquidity,
  receivedPoolFees,
  receivedPoolVolume,
  receivedPoolLiquidity,
  receivedPoolTxCount,
  receivedPoolAPY,
  receivedPoolTransactions,
} from "./actions"

import {
  APYChart,
  FeeVolumeChart,
  LiquidityChart,
  TxCountChart,
  VolumeChart,
} from "Data/Chart"
import { FetchDecodeError, FetchDecodeResult } from "Data/FetchDecode"
import * as Transaction from "Data/Transaction"

export type State = {
  aggLiquidity: RemoteData<FetchDecodeError, LiquidityChart.Type>
  aggVolume: RemoteData<FetchDecodeError, VolumeChart.Type>
  poolFees: RemoteData<FetchDecodeError, FeeVolumeChart.Type>
  poolVolume: RemoteData<FetchDecodeError, VolumeChart.Type>
  poolLiquidity: RemoteData<FetchDecodeError, LiquidityChart.Type>
  poolTxCount: RemoteData<FetchDecodeError, TxCountChart.Type>
  poolAPY: RemoteData<FetchDecodeError, APYChart.Type>
  poolTransactions: RemoteData<FetchDecodeError, Transaction.WithTotalValue[]>
}

export const initialState: State = {
  aggLiquidity: pending,
  aggVolume: pending,
  poolFees: pending,
  poolVolume: pending,
  poolLiquidity: pending,
  poolTxCount: pending,
  poolAPY: pending,
  poolTransactions: pending,
}

// side-effectful: function that logs on Left, and returns the typical fromEither
// TODO: move this to chain with the decoder stuff
function fromEitherLogFail<A>({
  payload,
}: PayloadAction<FetchDecodeResult<A>>): RemoteData<FetchDecodeError, A> {
  if (E.isLeft(payload)) {
    console.error("Fetch error:", payload.left)
  }
  return fromEither(payload)
}

export default createReducer(
  initialState,
  (builder: ActionReducerMapBuilder<State>) =>
    builder
      .addCase(receivedAggLiquidity, (state, action) => ({
        ...state,
        aggLiquidity: fromEitherLogFail(action),
      }))
      .addCase(receivedAggVolume, (state, action) => ({
        ...state,
        aggVolume: fromEitherLogFail(action),
      }))
      .addCase(receivedPoolFees, (state, action) => ({
        ...state,
        poolFees: fromEitherLogFail(action),
      }))
      .addCase(receivedPoolVolume, (state, action) => ({
        ...state,
        poolVolume: fromEitherLogFail(action),
      }))
      .addCase(receivedPoolLiquidity, (state, action) => ({
        ...state,
        poolLiquidity: fromEitherLogFail(action),
      }))
      .addCase(receivedPoolTxCount, (state, action) => ({
        ...state,
        poolTxCount: fromEitherLogFail(action),
      }))
      .addCase(receivedPoolAPY, (state, action) => ({
        ...state,
        poolAPY: fromEitherLogFail(action),
      }))
      .addCase(receivedPoolTransactions, (state, action) => ({
        ...state,
        poolTransactions: fromEitherLogFail(action),
      }))
)
