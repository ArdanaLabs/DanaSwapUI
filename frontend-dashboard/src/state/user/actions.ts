import { createAction } from "@reduxjs/toolkit"

import * as O from "fp-ts/Option"

import { SupportedW3ColorScheme, Theme } from "Data/User/Theme"

export const updateUserTheme = createAction<Theme>("user/updateUserTheme")

export const updatePrefersColorScheme = createAction<
  O.Option<SupportedW3ColorScheme>
>("user/updatePrefersColorScheme")
