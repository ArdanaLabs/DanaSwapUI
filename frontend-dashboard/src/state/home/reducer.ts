import {
  ActionReducerMapBuilder,
  PayloadAction,
  createReducer,
} from "@reduxjs/toolkit"

import { RemoteData, fromEither, pending } from "fp-ts-remote-data"

import { CombinedStats } from "Data/Stats/CombinedStats"
import { FetchDecodeError, FetchDecodeResult } from "Data/FetchDecode"

import { receivedCombinedStats } from "./actions"

export type State = RemoteData<FetchDecodeError, CombinedStats>

export const initialState = pending

export default createReducer(
  initialState,
  (builder: ActionReducerMapBuilder<State>) =>
    builder.addCase(
      receivedCombinedStats,
      (
        _state: State,
        action: PayloadAction<FetchDecodeResult<CombinedStats>>
      ) => fromEither(action.payload)
    )
)
