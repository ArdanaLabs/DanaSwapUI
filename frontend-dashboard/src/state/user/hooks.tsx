// import { useCallback } from 'react';
import { useDispatch, useSelector } from "react-redux"

import * as O from "fp-ts/Option"
import { pipe } from "fp-ts/function"

import * as Theme from "Data/User/Theme"

import { AppDispatch, AppState } from "state"
import { UserState } from "./reducer"
import { updateUserTheme } from "./actions"

export function getTheme(
  userState: Pick<UserState, "theme" | "prefersColorScheme">
): Theme.Theme {
  return pipe(
    userState.theme,
    O.alt(
      (): O.Option<Theme.Theme> =>
        O.chain(
          (s: Theme.SupportedW3ColorScheme): O.Option<Theme.Theme> =>
            Theme.fromW3ColorScheme(s)
        )(userState.prefersColorScheme)
    ),
    O.getOrElse((): Theme.Theme => Theme.defaultTheme)
  )
}

export function useUserTheme(): Theme.Theme {
  const user: UserState = useSelector<AppState, UserState>(
    (state) => state.user
  )
  return getTheme(user)
}

export function useUserThemeManager(): [
  Theme.Theme,
  (newTheme: Theme.Theme) => void
] {
  const dispatch = useDispatch<AppDispatch>()
  const theme: Theme.Theme = useUserTheme()

  // const setDarkMode = useCallback(
  //   (darkMode: boolean) => {
  //     dispatch(updateUserDarkMode({ userDarkMode: darkMode }));
  //   },
  //   [dispatch]
  // );
  const setTheme = (newTheme: Theme.Theme): void => {
    dispatch(updateUserTheme(newTheme))
  }

  return [theme, setTheme]
}
