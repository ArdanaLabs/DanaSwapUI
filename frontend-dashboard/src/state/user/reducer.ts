import { ActionReducerMapBuilder, createReducer } from "@reduxjs/toolkit"

import * as O from "fp-ts/Option"

import { SupportedW3ColorScheme, Theme } from "Data/User/Theme"

import { updatePrefersColorScheme, updateUserTheme } from "./actions"

function currentTimestamp(): number {
  return new Date().getTime()
}

export const THEME_LOCALSTORAGE_KEY = "user/theme"

export interface UserState {
  lastUpdateVersionTimestamp?: number
  theme: O.Option<Theme> // current theme
  prefersColorScheme: O.Option<SupportedW3ColorScheme>
  timestamp: number
}

export const initialState: UserState = {
  theme: O.none,
  prefersColorScheme: O.none,
  timestamp: currentTimestamp(), // side-effect
}

export default createReducer(
  initialState,
  (builder: ActionReducerMapBuilder<UserState>) =>
    builder
      .addCase(updateUserTheme, (state, action) => {
        localStorage.setItem(
          THEME_LOCALSTORAGE_KEY,
          action.payload.toString(10)
        )
        const timestamp = currentTimestamp()
        return { ...state, theme: O.some(action.payload), timestamp }
      })
      .addCase(updatePrefersColorScheme, (state, action) => {
        const timestamp = currentTimestamp()
        return { ...state, prefersColorScheme: action.payload, timestamp }
      })
)
