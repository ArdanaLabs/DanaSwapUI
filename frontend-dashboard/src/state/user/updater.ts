import { useEffect } from "react"
import { useDispatch } from "react-redux"

import * as O from "fp-ts/Option"
import { pipe } from "fp-ts/function"

import { Theme, SupportedW3ColorScheme } from "Data/User/Theme"

import { AppDispatch } from "state"
import { updatePrefersColorScheme, updateUserTheme } from "./actions"
import { THEME_LOCALSTORAGE_KEY } from "./reducer"

export default function Updater(): null {
  const dispatch = useDispatch<AppDispatch>()

  useEffect(() => {
    const storedTheme: O.Option<Theme> = pipe(
      localStorage.getItem(THEME_LOCALSTORAGE_KEY),
      O.fromNullable,
      O.chain((s: string): O.Option<Theme> => {
        const i: number = parseInt(s, 10)
        if (Number.isNaN(i) || !(i in Theme)) {
          return O.none
        } else {
          return O.some(i)
        }
      })
    )

    if (O.isSome(storedTheme)) {
      dispatch(updateUserTheme(storedTheme.value))
    }

    // This matching is done in this manner because the W3 spec clearly states
    // that we should expect the current set scheme support to be extended.
    // However, there’s not an easy way to just ‘get’ the string value, unlike
    // `devicePixelRatio`, so getting this match isn’t ‘easy’.

    const mkMatcher = (scheme: SupportedW3ColorScheme): MediaQueryList =>
      window.matchMedia(`screen and (prefers-color-scheme: ${scheme})`)

    const lightMatch: MediaQueryList = mkMatcher(SupportedW3ColorScheme.light)
    const darkMatch: MediaQueryList = mkMatcher(SupportedW3ColorScheme.dark)

    function matchHandler(_evt?: MediaQueryListEvent) {
      let scheme: O.Option<SupportedW3ColorScheme> = O.none
      if (lightMatch.matches) {
        scheme = O.some(SupportedW3ColorScheme.light)
      } else if (darkMatch.matches) {
        scheme = O.some(SupportedW3ColorScheme.dark)
      }
      dispatch(updatePrefersColorScheme(scheme))
    }

    matchHandler()

    lightMatch.addEventListener("change", matchHandler)
    darkMatch.addEventListener("change", matchHandler)

    return () => {
      lightMatch.removeEventListener("change", matchHandler)
      darkMatch.removeEventListener("change", matchHandler)
    }
  }, [dispatch])

  return null
}
