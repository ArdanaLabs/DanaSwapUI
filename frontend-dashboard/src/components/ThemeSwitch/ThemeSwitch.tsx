import React from "react"
import cx from "classnames"
import { Box } from "@material-ui/core"
import { makeStyles } from "@material-ui/core/styles"

import { useDarkModeManager } from "state/user/hooks"

import ICO_Dark from "assets/svg/moon.svg"
import ICO_Light from "assets/svg/sun.svg"

const useStyles = makeStyles(({ palette }) => ({
  self: {
    margin: "0 10px",
    padding: 5,
    cursor: "pointer",
    borderRadius: "100px",
    display: "flex",
    alignItems: "center",
    width: "175px",
    background: palette.primary.light,
  },
  switchIcon: {
    "borderRadius": "50%",
    "background": "#FFFFFF",
    "padding": 5,
    "width": 28,
    "height": 28,
    "display": "flex",
    "justifyContent": "center",
    "alignItems": "center",

    "& img": {
      width: "14px",
      height: "14px",
    },
  },
  switchLabel: {
    fontSize: "12px",
    fontWeight: 700,
    fontFamily: "Museo Sans",
    lineHeight: "16px",
    color: "#FFFFFF",
    flexGrow: 5,
    textAlign: "center",
  },
}))

const ThemeSwitch: React.FC = () => {
  const [darkMode, setDarkMode] = useDarkModeManager()
  const classes = useStyles()

  const toggleMode = () => {
    setDarkMode(!darkMode)
  }

  return (
    <Box
      className={cx(classes.self)}
      flexDirection={!darkMode ? "row" : "row-reverse"}
      onClick={toggleMode}
    >
      <Box className={cx(classes.switchLabel)}>
        {!darkMode ? "DARK MODE" : "LIGHT MODE"}
      </Box>
      <Box className={cx(classes.switchIcon)}>
        <img src={!darkMode ? ICO_Dark : ICO_Light} alt="Theme switch icon" />
      </Box>
    </Box>
  )
}

export default ThemeSwitch
