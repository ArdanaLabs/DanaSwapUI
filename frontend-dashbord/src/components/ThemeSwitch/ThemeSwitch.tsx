import React from 'react';
import { Typography, Box, useMediaQuery } from '@material-ui/core';
import { makeStyles, useTheme } from '@material-ui/core/styles';

import { useDarkModeManager } from 'state/user/hooks';
import { useDeviceWidth } from 'hooks';

import { SwitchWithGlider } from 'components';

const useStyles = makeStyles(({ palette }) => ({
  activeMode: {
    borderRadius: 10,
    '& svg': {
      marginRight: 8
    },
    '& svg path': {
      fill: palette.primary.main
    },
    '& .MuiTypography-root': {
      fontWeight: 700,
      fontSize: '14px',
      color: palette.primary.main
    }
  },
  inactiveMode: {
    backgroundColor: 'transparent',
    cursor: 'pointer',
    '& svg': {
      marginRight: 8
    },
    '& svg path': {
      fill: palette.secondary.main
    },
    '& .MuiTypography-root': {
      fontWeight: 400,
      fontSize: '14px',
      color: palette.secondary.main
    },
    '&:hover': {
      '& svg path': {
        fill: palette.text.primary
      },
      '& .MuiTypography-root': {
        fontWeight: 400,
        fontSize: '14px',
        color: palette.text.primary
      }
    }
  }
}));

const ThemeSwitch: React.FC = () => {
  const [darkMode, setDarkMode] = useDarkModeManager();
  const classes = useStyles();
  const theme = useTheme();
  const deviceWidth = useDeviceWidth();
  const { palette } = theme;
  const mobile = useMediaQuery(theme.breakpoints.down('sm'));

  const handleDayClick = () => {
    setTimeout(() => {
      setDarkMode(false);
    }, 0);
  };

  const handleNightClick = () => {
    setTimeout(() => {
      setDarkMode(true);
    }, 0);
  };

  const DayButton = () => (
    <Box
      display="flex"
      alignItems="center"
      justifyContent="center"
      className={!darkMode ? classes.activeMode : classes.inactiveMode}
      width={!mobile ? '80px' : deviceWidth / 2 - 16}
      height={!mobile ? '30px' : '36px'}
      onClick={handleDayClick}
    >
      <Typography>Day</Typography>
    </Box>
  );

  const NightButton = () => (
    <Box
      display="flex"
      width={!mobile ? '80px' : deviceWidth / 2 - 16}
      height={!mobile ? '30px' : '36px'}
      alignItems="center"
      justifyContent="center"
      className={darkMode ? classes.activeMode : classes.inactiveMode}
      onClick={handleNightClick}
    >
      <Typography>Night</Typography>
    </Box>
  );

  return (
    <Box
      display="flex"
      padding={!mobile ? '0 6px' : '0'}
      justifyContent="space-between"
      style={{ backgroundColor: palette.background.paper }}
    >
      {!mobile ? (
        <SwitchWithGlider
          elements={[DayButton, NightButton]}
          defaultIndex={!darkMode ? 0 : 1}
          marginBetweenSwitches={7}
          gliderWidth={80}
          gliderHeight={30}
        />
      ) : (
        <SwitchWithGlider
          elements={[DayButton, NightButton]}
          defaultIndex={!darkMode ? 0 : 1}
          gliderWidth={deviceWidth / 2 - 16}
          marginBetweenSwitches={8}
          gliderHeight={36}
        />
      )}
    </Box>
  );
};

export default ThemeSwitch;
