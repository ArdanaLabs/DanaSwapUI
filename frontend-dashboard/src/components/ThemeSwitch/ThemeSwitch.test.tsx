import Adapter from "@wojtekmaj/enzyme-adapter-react-17"
import * as Enzyme from "enzyme"
import { shallow } from "enzyme"

import * as O from "fp-ts/Option"

import { UserState } from "state/user/reducer"
import ThemeSwitch from "."

Enzyme.configure({ adapter: new Adapter() })

const mockUserState: UserState = {
  theme: O.none,
  prefersColorScheme: O.none,
  timestamp: 0,
}

jest.mock("react-redux", () => ({
  useDispatch: () => {},
  useSelector: (): UserState => mockUserState,
}))

let wrapper: Enzyme.ShallowWrapper

describe("Components / ThemeSwitch / ThemeSwitch", () => {
  beforeEach(() => {
    wrapper = shallow(<ThemeSwitch />)
  })

  it("renders", () => {
    expect(wrapper).toMatchSnapshot()
  })

  it("should contain one image", () => {
    expect(wrapper.find("img")).toHaveLength(1)
  })
})
