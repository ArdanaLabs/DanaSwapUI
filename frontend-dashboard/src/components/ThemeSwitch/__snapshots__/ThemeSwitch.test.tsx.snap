// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components / ThemeSwitch / ThemeSwitch renders 1`] = `ShallowWrapper {}`;
